### PR TITLE
lib/types: make pattern of strMatching accessible

### PR DESCRIPTION
**Pull request by hsjobeki**
Makes `pattern` of `strMatching` accessible in the option tree. 

Thus this allows to generate for example json-schema validation
https://json-schema.org/understanding-json-schema/reference/regular_expressions

Context: I am building a Frontend that is partially derived from those options. Exposing the pattern is very helpfull in that case.
Otherwise my users get evalModules eval stacktraces. Which is after the field was submitted / sent to the server.
If i knew the pattern upfront then i can display feedback as the user types. 

## Things  done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: lib/types.nix [hunk 1]
```diff
@@ -443,6 +443,10 @@ rec {
       descriptionClass = "noun";
       check = x: str.check x && builtins.match pattern x != null;
       inherit (str) merge;
+      functor = defaultFunctor "strMatching" // {
+        payload = pattern;
```
Comment (roberth) [inline lib/types.nix:447]: Quick thing for forward compatibility:

```suggestion
        payload = { inherit pattern; };
```

**Review by roberth — COMMENTED** [review]
I'm pretty sure this wasn't the original intent of the `functor` attribute.
We'll need to either encourage this or come up with an alternative.
@infinisil wdyt?

**Comment by hsjobeki** [comment]
@infinisil. I thought that it could make sense to move away from using `wrapped` in `defaultFunctor` entirely, for all types.

Instead using only `payload` with `payload.wrapped` beeing the new convention for wrapped.
I hoped this could reduce technical complexity as shown:

```nix
# Example for 'types.attrsOf'
functor = (defaultFunctor name) // {
        payload.wrapped = elemType;
        type = payload: attrsOf payload.wrapped;
        binOp = lhs: rhs: { 
           wrapped = lhs.wrapped.typeMerge rhs.wrapped.functor; 
           # Place for other payload merging
         };
      };
 ```
 
 Once all types use `payload` exclusively. This reduces complexity in the typeMerging functions.
 Because `payload` can include all of `wrapped` functionallity as well, and is extensible for type specific attributes.
 
 ```
   defaultTypeMerge = f: f':
    let
        payload = f.binOp f.payload f'.payload;
    in
    # cannot merge different types
    if f.name != f'.name
       then null
    # simple types
    else if  (f.payload == null && f'.payload == null)
       then f.type
    # all other types
    else if (f.payload != null && f'.payload != null)
       then f.type payload
    else null;
  ```

**Review by infinisil — APPROVED** [review]
After https://github.com/NixOS/nixpkgs/pull/363565, we could make `functor.payload` stable (along with `functor.{binOp,type}`, and kind of `functor.name`), but that's a bit of a separate discussion.

While this doesn't fix a problem related to Nixpkgs (seemingly nobody has ever needed type merging for `strMatching`!), this doesn't break anything and unblocks @hsjobeki with getting access to the pattern for a UI, so I think this is fine to merge :)

**Pull request merged**